src/Renci.SshNet/Security/Cryptography/Ciphers/AesGcmCipher.BouncyCastleImpl.cs

```diff
@@ -0,0 +1,52 @@
+﻿using Org.BouncyCastle.Crypto;
+using Org.BouncyCastle.Crypto.Engines;
+using Org.BouncyCastle.Crypto.Modes;
+using Org.BouncyCastle.Crypto.Parameters;
+
+using Renci.SshNet.Common;
+using Renci.SshNet.Messages.Transport;
+
+namespace Renci.SshNet.Security.Cryptography.Ciphers
+{
+    internal partial class AesGcmCipher
+    {
+        private sealed class BouncyCastleImpl : Impl
+        {
+            private readonly byte[] _key;
+            private readonly int _tagSize;
+            private readonly GcmBlockCipher _cipher;
+
+            public BouncyCastleImpl(byte[] key, int tagSize)
+            {
+                _key = key;
+                _tagSize = tagSize;
+                _cipher = new GcmBlockCipher(new AesEngine());
+            }
+
+            public override void Encrypt(byte[] nonce, byte[] input, int plainTextOffset, int plainTextLength, int associatedDataOffset, int associatedDataLength, byte[] output, int cipherTextOffset)
+            {
+                var parameters = new AeadParameters(new KeyParameter(_key), _tagSize * 8, nonce, input.Take(associatedDataOffset, associatedDataLength));
+                _cipher.Init(forEncryption: true, parameters);
```

Comment:
Seems a shame we need to initialise new parameters for each packet. Can the `KeyParameter` argument be a field? And is there a way to use `GcmCipher.ProcessAadBytes` to avoid the copy of the associated data?

Reply:
This applies to the line above (`new AeadParameters(...`)

Reply:
> And is there a way to use `GcmCipher.ProcessAadBytes` to avoid the copy of the associated data?
The limitation comes from BouncyCastle. The `AeadParameters`'s constructor doesn't take associated data with offset and length.

Reply:
Oh I see what you mean. Yes, we can use `GcmCipher.ProcessAadBytes`